### PR TITLE
Host Volumes Support: Rollup Edition

### PR DESCRIPTION
**Pull request by endocrimes**
This PR is a rollup PR for Host Volumes Support. All code has been reviewed in separate PRs and will eventually be merged into master from here.

## PRs:
- #5923: Initial Support For Host Volumes
- #6011: Host Volume support in `nomad node status`
- #6014: Add HostVolume ACLs

### CODE REVIEW

File: api/tasks.go [hunk 1]
```diff
@@ -362,6 +362,21 @@ func (m *MigrateStrategy) Copy() *MigrateStrategy {
 	return nm
 }
 
+type Volume struct {
```
Comment (notnoop) [inline api/tasks.go]: I'd urge us to have better code documentation for the significant data structures in this PR before merging, for our own sanity when we approach this code.  This is a follow up from https://github.com/hashicorp/nomad/pull/5923#discussion_r309283098

File: client/config/config.go [hunk 2]
```diff
@@ -234,6 +234,9 @@ type Config struct {
 	// for allocations in bridge networking mode. Subnet must be in CIDR
 	// notation
 	BridgeNetworkAllocSubnet string
+
+	// HostVolumes is the set of configured host volumes
+	HostVolumes map[string]*structs.ClientHostVolumeConfig
```
Comment (notnoop) [inline client/config/config.go]: Document the mapping here.  What's the map key?

File: jobspec/parse_group.go [hunk 3]
```diff
@@ -274,3 +283,61 @@ func parseRestartPolicy(final **api.RestartPolicy, list *ast.ObjectList) error {
...
+		if err := helper.CheckHCLKeys(item.Val, valid); err != nil {
+			return err
+		}
+
+		var m map[string]interface{}
+		if err := hcl.DecodeObject(&m, item.Val); err != nil {
+			return err
+		}
+
+		// TODO(dani): FIXME: this is gross but we don't have ObjectList.Filter here
```
Comment (notnoop) [inline jobspec/parse_group.go]: If this is not a bug and it isn't going to be addressed before merging, I'd suggest dropping the `FIXME` label.

File: nomad/job_endpoint.go [hunk 4]
```diff
@@ -89,6 +89,24 @@ func (j *Job) Register(args *structs.JobRegisterRequest, reply *structs.JobRegis
 		if !aclObj.AllowNsOp(args.RequestNamespace(), acl.NamespaceCapabilitySubmitJob) {
 			return structs.ErrPermissionDenied
 		}
+		// Validate Volume Permsissions
+		for _, tg := range args.Job.TaskGroups {
+			for _, vol := range tg.Volumes {
+				if vol.Type != structs.VolumeTypeHost {
+					return structs.ErrPermissionDenied
```
Comment (notnoop) [inline nomad/job_endpoint.go]: not a big deal but this seems more invalid request or unsupported type than a permission error, same as the parsing error below.

File: api/tasks.go (hunk 1 above)
Comment (endocrimes) [reply api/tasks.go]: @notnoop The structs were documented in `nomad/structs/volumes.go` already, I've copied the documentation across to here, but our api structs documentation is generally pretty nonexistent.

File: nomad/job_endpoint.go (hunk 4 above)
Comment (endocrimes) [reply nomad/job_endpoint.go]: Yep - it'll actually be caught earlier in the process, (validation of the submission happens before permissions validation), this is entirely being overly defensive.

**Pull request merged**